lib/server.js

```diff
@@ -826,12 +825,32 @@ Server.prototype._run = function _run(req, res, route, chain, cb) {
 
         if (arg) {
             if (arg instanceof Error) {
-                errName = arg.name.replace(/Error$/, '');
-                log.trace({err: arg, errName: errName}, 'next(err=%s)',
-                    (arg.name || 'Error'));
 
+                var errName = arg.name.replace(/Error$/, '');
+                log.trace({
+                    err: arg,
+                    errName: errName
+                }, 'next(err=%s)', (arg.name || 'Error'));
+
+                // always attempt to use the most specific error listener
+                // possible. fall back on generic 'error' listener if we can't
+                // find one for the error we got.
+                var hasErrListeners = false;
+                var errEvtName;
+
+                // if we have listeners for the specific error
                 if (self.listeners(errName).length > 0) {
-                    self.emit(errName, req, res, arg, once(function () {
+                    hasErrListeners = true;
+                    errEvtName = errName;
+                }
+                // or if we have a generic error listener
+                else if (self.listeners('error').length > 0) {
```

Comment:
As envisioned originally we'd only fire one event, whatever was the most specific match found. 

The generic error event would only ever fire if you didn't register a specific handler for that error type. Thinking about it now, it seems more reasonable to _always_ fire the generic event. 

I think we should fire the most specific one first, then the generic listener. Sound reasonable?